browser/src/view.js

```diff
@@ -0,0 +1,511 @@
+    document.getElementById('channel').textContent = this.model.channel;
+    document.getElementById('zoom').textContent = `${this.canvas.zoom}%`;
+
+    const minX = Math.floor(this.canvas.sx);
+    const maxX = Math.ceil(this.canvas.sx + this.canvas.sWidth);
+    const displayedX = `${minX}-${maxX}`;
+    document.getElementById('displayedX').textContent = displayedX;
+
+    const minY = Math.floor(this.canvas.sy);
+    const maxY = Math.ceil(this.canvas.sy + this.canvas.sHeight);
+    const displayedY = `${minY}-${maxY}`;
+    document.getElementById('displayedY').textContent = displayedY;
+
+    const highlightText = (this.model.highlight) ? 'ON' : 'OFF';
+    document.getElementById('highlight').textContent = highlightText;
+    document.getElementById('edit_brush').textContent = this.model.size;
+    this.renderLabelRows();
+    // always show tool and selected labels
+
+    document.getElementById('mode').textContent = this.getTool();
+    const foreground = this.model.foreground;
+    const background = this.model.background;
+    document.getElementById('foreground').textContent = foreground === 0 ? 'background' : foreground;
+    document.getElementById('background').textContent = background === 0 ? 'background' : background;
+
+    this.renderPrompt();
+  }
+
+  getTool() {
+    const states = window.controller.service.state.toStrings();
+    const toolbarState = 'mouse.toolbar.';
+    for (const state of states) {
+      if (state.startsWith(toolbarState)) { return state.substring(toolbarState.length); }
+    }
+    return 'loading...';
+  }
+
+  /**
+   * Renders the rows about the label being hovered over.
+   */
+  renderLabelRows() {
+    if (this.canvas.label !== 0) {
+      document.getElementById('label').textContent = this.canvas.label;
+      const track = this.model.tracks[this.model.feature][this.canvas.label.toString()];
+      document.getElementById('slices').textContent = track.slices.toString();
+    } else {
+      document.getElementById('label').textContent = '';
+      document.getElementById('slices').textContent = '';
+    }
+  }
+
+  renderPrompt() {
+    const state = window.controller.service.state;
+    const label = this.model.foreground;
+    const secondLabel = this.model.background;
+    const frame = this.model.frame;
+    let prompt;
+    if (state.matches('confirm.predictFrame')) {
+      prompt = `Predict labels on frame ${frame}?`;
+    } else if (state.matches('confirm.predictAll')) {
+      prompt = 'Predict labels on ALL frames?';
+    } else if (state.matches('confirm.replaceFrame')) {
+      prompt = `Replace label ${secondLabel} with label ${label} on frame ${frame}?`
+    } else if (state.matches('confirm.replaceAll')) {
+      prompt = `Replace label ${secondLabel} with label ${label} on ALL frames?`
+    } else if (state.matches('confirm.swapFrame')) {
+      prompt = `Swap label ${label} with label ${secondLabel} on frame ${frame}?`
+    }
```

Comment:
Can these states be defined in the state chart? Or is that not recommended?

Reply:
Not sure what you mean by defining these states on the state chart. All of these states (`confirm.predictFrame`, `confirm.predictAll`, etc) are already in the state chart, and we're checking if we are in that state here because we need to display different prompts depending on the state.

I tried storing the prompt on the state machine initially, but that caused some bugs because the prompt should update as the selected labels change, so putting them in the view lets us access the correct selected labels.

Reply:
Sorry, I meant storing the prompts as part of the state machine, but you've addressed that.